test/jdk/jdk/jfr/event/oldobject/TestFieldInformation.java

```diff
@@ -51,23 +51,37 @@ public class TestFieldInformation {
     public static void main(String[] args) throws Exception {
         WhiteBox.setWriteAllObjectSamples(true);
 
+        // OldObjectSample is emitted at the end of the recording, and we need
+        // to let GC catch up before we do the event. There is no reason to wait
```

Comment:
This notion of a "GC catching up" confuses these tests. Why should the GC " catch up" before a recording is stopped? If you need to use this construct here, please specify that it applies only to Shenandoah GC and not to GCs in general.

Reply:
This applies to any concurrent GC, really.

We have allocated things, GC might have been triggered, we need to wait for "old objects" to appear now. Current tests assume that "old" objects appear "instantaneously" after the allocations -- which is somewhat sane for STW collectors, but not for the concurrent ones. The fact that Shenandoah does not report old objects while GC is running is just a facet of this general problem.

Reply:
Sampled oops are referenced by weak handles in the JFR sample priority queue. When the recording stops, an exclusive safe point operation occurs that dumps the ObjectSamples in the priority queue and possibly walks the paths-to-gc-roots. Can you explain more what you mean by "wait for old objects to appear"? Appear where?

Reply:
All right, I think I misunderstood how sampling works. I assumed it happens from the GC cycle, but now I see it happens from the allocation path. Indeed sampler holds the object in weak oop storage, and they are supposed to be accessible immediately after the allocation. 

I tightened up comments to reflect this understanding. Looks better, or?

Reply:
I don't know. The backoff mechanism now introduced is to handle the special case where the LeakProfiler dump operation, even though it is performed under an exclusive safepoint, happens to to coincide with some phase of Shenandoah GC - a lot of surface for handling this special case.

Also, need this not be categorical for all OldObject tests?

Reply:
Yes. I think we go for a more comprehensive fix: https://github.com/openjdk/jdk/pull/20328#issuecomment-2434920739